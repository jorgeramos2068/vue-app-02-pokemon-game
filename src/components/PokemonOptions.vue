<template>
  <div class="options-container">
    <ul>
      <li>Opt1</li>
      <li>Opt2</li>
      <li>Opt3</li>
      <li>Opt4</li>
    </ul>
  </div>
</template>

<script>
export default {
  name: 'PokemonOptions',
};
</script>

<style scoped>
ul {
  list-style-type: none;
  margin-top: 50px;
}
li {
  background-color: white;
  border-radius: 5px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  cursor: pointer;
  margin-bottom: 20px;
  padding: 5px 0;
  width: 250px;
}

li:hover {
  background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
  display: flex;
  justify-content: center;
}
</style>
