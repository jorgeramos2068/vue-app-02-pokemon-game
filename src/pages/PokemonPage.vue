<template>
  <div>
    <div v-if="!pokemon">
      <h1>Loading...</h1>
    </div>
    <div v-else>
      <h1>Who is this Pokemon?</h1>
      <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
      <PokemonOptions
        :pokemonOptions="pokemonOptions"
        @selection="checkAnswer"
      />
      <template v-if="showMessage">
        <h2 class="fade-in">{{ message }}</h2>
        <button @click="newGame()">New game</button>
      </template>
    </div>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue';
import PokemonOptions from '@/components/PokemonOptions.vue';
import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
  name: 'PokemonPage',
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonOptions: [],
      pokemon: null,
      showPokemon: false,
      showMessage: false,
      message: '',
    };
  },
  methods: {
    async getPokemons() {
      this.pokemonOptions = await getPokemonOptions();
      const randomInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonOptions[randomInt];
    },
    checkAnswer(selectedId) {
      this.showPokemon = true;
      this.showMessage = true;
      if (selectedId === this.pokemon.id) {
        this.message = `Correct, ${this.pokemon.name}`;
      } else {
        this.message = `Oops, the correct answer is ${this.pokemon.name}`;
      }
    },
    newGame() {
      this.pokemonOptions = [];
      this.pokemon = null;
      this.showPokemon = false;
      this.showMessage = false;
      this.message = '';
      this.getPokemons();
    },
  },
  mounted() {
    this.getPokemons();
  },
};
</script>
