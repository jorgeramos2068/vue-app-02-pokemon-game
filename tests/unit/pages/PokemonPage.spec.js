import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import PokemonPicture from '@/components/PokemonPicture';
import PokemonOptions from '@/components/PokemonOptions';
import { mockPokemons } from '../__mocks__/mockPokemons';

describe('PokemonPage component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  afterAll(done => {
    done();
  });

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match the snapshot when the data is loaded', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonOptions: mockPokemons,
          pokemon: mockPokemons[0],
          showPokemon: false,
          showMessage: false,
          message: '',
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call getPokemons method on mounting', () => {
    const getPokemonsSpy = jest.spyOn(PokemonPage.methods, 'getPokemons');
    shallowMount(PokemonPage);
    expect(getPokemonsSpy).toHaveBeenCalled();
  });

  it('should show PokemonPicture and PokemonOptions components', async () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonOptions: mockPokemons,
          pokemon: mockPokemons[0],
          showPokemon: false,
          showMessage: false,
          message: '',
        };
      },
    });
    const pokemonPicture = wrapper.findComponent(PokemonPicture);
    expect(pokemonPicture.exists()).toBe(true);
    const pokemonOptions = wrapper.findComponent(PokemonOptions);
    expect(pokemonOptions.exists()).toBe(true);
    expect(pokemonPicture.vm.pokemonId).toBe(mockPokemons[0].id);
    expect(pokemonOptions.vm.pokemonOptions).toEqual(mockPokemons);
  });

  it('should check the checkAnswer method', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonOptions: mockPokemons,
          pokemon: mockPokemons[0],
          showPokemon: false,
          showMessage: false,
          message: '',
        };
      },
    });
    await wrapper.vm.checkAnswer(mockPokemons[0].id);
    expect(wrapper.find('h2').exists()).toBeTruthy();
    expect(wrapper.find('h2').text()).toBe(`Correct, ${mockPokemons[0].name}`);
    expect(wrapper.vm.showPokemon).toBeTruthy();
    await wrapper.vm.checkAnswer(mockPokemons[1].id);
    expect(wrapper.vm.message).toBe(
      `Oops, the correct answer is ${mockPokemons[0].name}`
    );
  });
});
