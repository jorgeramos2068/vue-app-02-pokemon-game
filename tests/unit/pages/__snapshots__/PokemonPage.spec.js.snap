// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component should match the snapshot 1`] = `
<div>
  <div>
    <h1>Loading...</h1>
  </div>
</div>
`;

exports[`PokemonPage component should match the snapshot when the data is loaded 1`] = `
<div>
  <div>
    <h1>Who is this Pokemon?</h1>
    <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
    <pokemon-options-stub pokemonoptions="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--v-if-->
  </div>
</div>
`;
